fairscale/experimental/nn/sync_batchnorm.py

```diff
@@ -62,6 +40,87 @@ def _forward(
     _forward = torch.jit.script(_forward)  # type: ignore
 
 
+class _SyncBatchNormFunction(torch.autograd.Function):
+    @staticmethod
+    # type: ignore
+    def forward(
+        ctx, input, weight, bias, affine, track_running_stats, running_mean, running_var, eps, momentum, process_group
+    ):
+        dim = [d for d in range(input.ndim) if d != 1]
+        count = torch.full((1,), input.numel() // input.size(1), device=input.device, dtype=input.dtype)
+        total_count = count.clone()
+        handle = dist.all_reduce(total_count, group=process_group, async_op=True)
+        mean = torch.mean(input, dim=dim, keepdim=True)
+        meansqr = torch.mean(input * input, dim=dim, keepdim=True)
+        vec = torch.cat([mean, meansqr])
+        handle.wait()
+        vec = vec * (count / total_count)
+        dist.all_reduce(vec, group=process_group)
+        mean, meansqr = vec.chunk(2)
+        var = meansqr - mean * mean
+        invstd = torch.rsqrt(var + eps)
+
+        ctx.save_for_backward(input, weight, bias, mean, invstd, total_count)
+        ctx.process_group = process_group
+
+        return _forward(
+            input,
+            affine,
+            track_running_stats,
+            mean,
+            var,
+            invstd,
+            momentum,
+            weight,
+            bias,
+            running_mean,
+            running_var,
+            total_count,
+        )
+
+    @staticmethod
+    # type: ignore
+    def backward(ctx, grad_output):
+        grad_input = None
+        grad_weight = None
+        grad_bias = None
+        input, weight, bias, mean, invstd, total_count = ctx.saved_tensors
+        process_group = ctx.process_group
+        dim = [d for d in range(input.ndim) if d != 1]
+        if ctx.needs_input_grad[0] or ctx.needs_input_grad[1]:
```

Comment:
since this needs_input_grad list seems to be maintain somewhere else, it is better to add a comment and document their meaning and whether one of them must be true or they can be both true or both false?

Reply:
will be useful for people making changes here in the future.

Reply:
Good idea. needs_input_grad is maintained by autograd function base class.